### PR TITLE
x86-64: Select REP MOVS instruction based on CPU for char arraycopy

### PR DESCRIPTION
**Pull request by a7ehuo**
Use `rep movsw` to copy multiple of 2 bytes for copy size greater than the threshold (32 bytes), which performs better than `rep movsd` for CascadeLake and later CPUs.

Setting env variable `TR_UseREPMOVSWFor16BitPrimitiveArrayCopy` forces to use `rep movsw`.

Setting env variable `TR_UseREPMOVSDFor16BitPrimitiveArrayCopy` forces to use `rep movsd`.

### CODE REVIEW

**Comment by a7ehuo** [comment]
@0xdaryl May I ask you to review this change? Thank you! 

@vijaysun-omr @hzongaro fyi

File: compiler/x/codegen/OMRTreeEvaluator.cpp [hunk 1]
```diff
@@ -2588,14 +2637,17 @@ static void arrayCopyPrimitiveInlineSmallSizeWithoutREPMOVS(TR::Node* node,
...
+      generateRepMovsInstructionBasedOnElementSize(elementSize, true /* basedOnCPU */, node, dstReg, srcReg, sizeReg, NULL, mainEndLabel, cg);
 
       {
       TR_OutlinedInstructionsGenerator og(backwardLabel, node, cg);
       generateRegMemInstruction(TR::InstOpCode::LEARegMem(), node, srcReg, generateX86MemoryReference(srcReg, sizeReg, 0, -(intptr_t)elementSize, cg), cg);
       generateRegMemInstruction(TR::InstOpCode::LEARegMem(), node, dstReg, generateX86MemoryReference(dstReg, sizeReg, 0, -(intptr_t)elementSize, cg), cg);
       generateInstruction(TR::InstOpCode::STD, node, cg);
-      generateRepMovsInstruction(repmovs, node, sizeReg, NULL, cg);
+
+      generateRepMovsInstructionBasedOnElementSize(elementSize, false /* basedOnCPU */, node, dstReg, srcReg, sizeReg, NULL, mainEndLabel, cg);
```
Comment (0xdaryl) [inline compiler/x/codegen/OMRTreeEvaluator.cpp:2649]: Why is this REP MOVS[x] not decided based on the CPU while the others are?

Comment (a7ehuo) [reply compiler/x/codegen/OMRTreeEvaluator.cpp:2649]: This is part of the backward copy in the out of line instructions, which are not within any internal control flow. Using `rep movsd` copying char array (element size = 2 bytes) introduces a jump instruction `jae`.  This is also aligned with the original implementation on backward copy that always uses `rep movsw` for backward copy when the forward copy uses `rep movsd`. 

Here is an example of running this change on Broadwell that chooses `rep movsd` for forward copy based on CPU and always uses `rep movsw` for backward copy as the original implementation. 

```
 [0x7ff590d7a060] Label L0048:         # label  # (Start of internal control flow)
 [0x7ff590d7a2e0] cmp   rcx, 0x00000010   # CMP8RegImm4
 [0x7ff590d7a370] ja Label L0053       # JA4
 [0x7ff590d7a410] cmp   rcx, 0x00000002   # CMP8RegImm4
 [0x7ff590d7a4a0] ja Label L0051       # JA4
 [0x7ff590d7a540] test  rcx, rcx    # TEST8RegReg
 [0x7ff590d7a5d0] je Label L0049       # JE4
 [0x7ff590d7a700] mov   ax, word ptr [rsi]      # L2RegMem
 [0x7ff590d7a820] mov   word ptr [rdi], ax      # S2MemReg
 [0x7ff590d7a8b0] jmp   Label L0049       # JMP4
 [0x7ff590d7a950] Label L0051:         # label
 [0x7ff590d7a9f0] cmp   rcx, 0x00000008   # CMP8RegImm4
 [0x7ff590d7aa80] ja Label L0052       # JA4
 [0x7ff590d7abb0] mov   eax, dword ptr [rsi+1*rcx-0x4]      # L4RegMem, SymRef [#442 -4]
 [0x7ff590d7acd0] mov   edx, dword ptr [rsi]    # L4RegMem
 [0x7ff590d7adf0] mov   dword ptr [rdi+1*rcx-0x4], eax      # S4MemReg, SymRef [#444 -4]
 [0x7ff590d7af10] mov   dword ptr [rdi], edx    # S4MemReg
 [0x7ff590d7afa0] jmp   Label L0049       # JMP4
 [0x7ff590d7b040] Label L0052:         # label
 [0x7ff590d7b170] mov   rax, qword ptr [rsi+1*rcx-0x8]      # L8RegMem, SymRef [#446 -8]
 [0x7ff590d7b290] mov   rdx, qword ptr [rsi]    # L8RegMem
 [0x7ff590d7b3b0] mov   qword ptr [rdi+1*rcx-0x8], rax      # S8MemReg, SymRef [#448 -8]
 [0x7ff590d7b4d0] mov   qword ptr [rdi], rdx    # S8MemReg
 [0x7ff590d7b560] jmp   Label L0049       # JMP4
 [0x7ff590d7b600] Label L0053:         # label
 [0x7ff590d7b6a0] cmp   rcx, 0x00000020   # CMP8RegImm4
 [0x7ff590d7b730] ja Label L0050       # JA4
 [0x7ff590d7b860] movdqu   xmm0, xmmword ptr [rsi+1*rcx-0x10]     # MOVDQURegReg, SymRef [#450 -16]
 [0x7ff590d7b980] movdqu   xmm1, xmmword ptr [rsi]    # MOVDQURegReg
 [0x7ff590d7baa0] movdqu   xmmword ptr [rdi+1*rcx-0x10], xmm0     # MOVDQUMemReg, SymRef [#452 -16]
 [0x7ff590d7bbc0] movdqu   xmmword ptr [rdi], xmm1    # MOVDQUMemReg
 [0x7ff590d7bc50] jmp   Label L0049       # JMP4
 [0x7ff590d7bcf0] Label L0050:         # label                    //<------ test copy direction during run time
 [0x7ff590d7bdf0] sub   rdi, rsi    # SUB8RegReg
 [0x7ff590d7be80] cmp   rdi, rcx    # CMP8RegReg
 [0x7ff590d7bfa0] lea   rdi, qword ptr [rdi+1*rsi]    # LEA8RegMem
 [0x7ff590d7f310] vfpSave        # AdjustFramePtr
 [0x7ff590d7c030] jb Outlined Label L0055       # JB4      // jump to backward copy
 [0x7ff590d7c0d0] shr   rcx, 0x02   # SHR8RegImm1
 [0x7ff590d7c160] rep movsd                       # REPMOVSD //<------ 
 [0x7ff590d7c1e0] jae   Label L0049       # JAE1                  //<------
 [0x7ff590d7c310] mov   cx, word ptr [rsi]      # L2RegMem
 [0x7ff590d7c430] mov   word ptr [rdi], cx      # S2MemReg
 [0x7ff590d7ce20] assocreg                        # assocreg
 [0x7ff590d7cbc0] Label L0049:         # label  # (End of internal control flow)
...
...
------------ start out-of-line instructions
                                     
 [0x7ff590d7c530] Outlined Label L0055:         # label
 [0x7ff590d7c660] lea   rsi, qword ptr [rsi+1*rcx-0x2]      # LEA8RegMem, SymRef [#457 -2]
 [0x7ff590d7c780] lea   rdi, qword ptr [rdi+1*rcx-0x2]      # LEA8RegMem, SymRef [#458 -2]
 [0x7ff590d7c810] std                             # STD
 [0x7ff590d7c890] shr   rcx         # SHR8Reg1
 [0x7ff590d7c920] rep movsw                       # REPMOVSW
 [0x7ff590d7c9a0] cld                             # CLD
 [0x7ff590d7ca20] jmp   Label L0049       # JMP4
 [0x7ff590d7cb20] Label L0056:         # label
------------ end out-of-line instructions
```

Comment (0xdaryl) [reply compiler/x/codegen/OMRTreeEvaluator.cpp:2649]: OK thanks, so it is a simplification of the logic for a cold case.

**Comment by 0xdaryl** [comment]
Jenkins build xlinux,xmac

**Review by 0xdaryl — APPROVED** [review]

**Pull request merged**